src/gems/simulation/extra_output.py

```diff
@@ -0,0 +1,155 @@
+from dataclasses import dataclass, field
+from typing import Any, Callable, Dict
+
+from gems.expression import evaluate
+from gems.expression.evaluate import EvaluationError, ValueProvider
+from gems.expression.expression import ExpressionNode
+from gems.study.data import ComponentParameterIndex, TimeScenarioIndex
+
+
+@dataclass
+class ExtraOutput:
+    """Stores evaluated outputs (from ExpressionNodes), not solver variables."""
+
+    name: str
+    values: Dict[TimeScenarioIndex, float] = field(default_factory=dict)
+
+    def set(self, t: int, s: int, value: float) -> None:
+        self.values[TimeScenarioIndex(t, s)] = value
+
+    def get(self, t: int, s: int) -> float | None:
+        return self.values.get(TimeScenarioIndex(t, s))
+
+
+def evaluate_all_extra_outputs(
+    problem: Any, component_lookup: Callable[[str], Any]
+) -> Dict[str, Dict[str, ExtraOutput]]:
+    """
+    Evaluate all model-defined extra outputs for every component in the network.
+    """
+    if problem is None or getattr(problem, "context", None) is None:
+        return {}
+
+    database = getattr(problem.context, "database", None)
+    if database is None:
+        print("[WARN] No database found in problem context; extra outputs skipped.")
+        return {}
+
+    results: Dict[str, Dict[str, ExtraOutput]] = {}
+
+    for cmp in problem.context.network.all_components:
+        model_def = cmp.model
+        if not model_def or not model_def.extra_outputs:
+            continue
```

Comment:
I prefer using the negated statement (`if model_def and model_def.extra_outputs`) and then perform extra-outputs computation under the if, rather than use `continue` which I find less readable